pkg/migration/migration.go

```diff
@@ -0,0 +1,145 @@
+package migration
+
+import (
+	"fmt"
+
+	"persistent-volume-migrator/pkg/k8sutil"
+	logger "persistent-volume-migrator/pkg/log"
+
+	v1 "k8s.io/api/core/v1"
+	k8s "k8s.io/client-go/kubernetes"
+)
+
+const (
+	pvcCreateTimeout = 5
+)
+
+func MigrateToCSI(kubeConfig, sourceStorageClass, destinationStorageClass, rookNamespace, cephClusterNamespace string) error {
```

Comment:
Can you share what the log looks like now for converting a volume during a manual test? Or does the CI capture the log for a conversion?

Reply:
We can catch all the updated logs in our CI in gh actions. See [Migrate-to-Flex job run](https://github.com/ceph/persistent-volume-migrator/runs/4234435429?check_suite_focus=true#step:8:37)

Reply:
Snippet of the logs:

```I1117 06:07:37.704882      58 log.go:34] Create Kubernetes Client
I1117 06:07:37.707076      58 log.go:34] List all the PVC from the source storageclass
I1117 06:07:37.792236      58 log.go:34] 1 PVCs found with source StorageClass rook-ceph-block 
I1117 06:07:37.792259      58 log.go:34] Start Migration of PVCs to CSI
I1117 06:07:37.792265      58 log.go:34] migrating PVC "rbd-pvc"
I1117 06:07:37.792270      58 log.go:34] Fetch PV information from PVC rbd-pvc
I1117 06:07:37.809539      58 log.go:34] PV found "pvc-11c4c863-b866-4cbc-8028-fe045f9e3bec" 
I1117 06:07:37.811228      58 log.go:34] Update Reclaim policy from Delete to Reclaim for PV: pvc-11c4c863-b866-4cbc-8028-fe045f9e3bec
I1117 06:07:37.822791      58 log.go:34] Retrieving old ceph volume name from PV object: pvc-11c4c863-b866-4cbc-8028-fe045f9e3bec
I1117 06:07:37.822863      58 log.go:34] Flex rbdImageName name pvc-11c4c863-b866-4cbc-8028-fe045f9e3bec 
I1117 06:07:37.822916      58 log.go:34] Deleting pvc object: rbd-pvc
waiting for PVC rbd-pvc in state &PersistentVolumeClaimStatus{Phase:Bound,AccessModes:[ReadWriteOnce],Capacity:ResourceList{storage: {{1073741824 0} {<nil>} 1Gi BinarySI},},Conditions:[]PersistentVolumeClaimCondition{},} to be deleted (0 seconds elapsed) 
waiting for PVC rbd-pvc in state &PersistentVolumeClaimStatus{Phase:Bound,AccessModes:[ReadWriteOnce],Capacity:ResourceList{storage: {{1073741824 0} {<nil>} 1Gi BinarySI},},Conditions:[]PersistentVolumeClaimCondition{},} to be deleted (2 seconds elapsed) 
I1117 06:07:39.837808      58 log.go:34] Generate new PVC with same name in destination storageclass
I1117 06:07:39.838606      58 log.go:34] Structure of the generated PVC in destination Storageclass &PersistentVolumeClaim{ObjectMeta:{rbd-pvc  default  11c4c863-b866-4cbc-8028-fe045f9e3bec  0 2021-11-17 06:06:11 +0000 UTC <nil> <nil> map[] map[] [] [kubernetes.io/pvc-protection]  [{kubectl-create Update v1 2021-11-17 06:06:11 +0000 UTC FieldsV1 {"f:spec":{"f:accessModes":{},"f:resources":{"f:requests":{".":{},"f:storage":{}}},"f:storageClassName":{},"f:volumeMode":{}}}} {kube-controller-manager Update v1 2021-11-17 06:06:27 +0000 UTC FieldsV1 {"f:metadata":{"f:annotations":{".":{},"f:pv.kubernetes.io/bind-completed":{},"f:pv.kubernetes.io/bound-by-controller":{},"f:volume.beta.kubernetes.io/storage-provisioner":{}}},"f:spec":{"f:volumeName":{}}}} {kube-controller-manager Update v1 2021-11-17 06:06:27 +0000 UTC FieldsV1 {"f:status":{"f:accessModes":{},"f:capacity":{".":{},"f:storage":{}},"f:phase":{}}}}]},Spec:PersistentVolumeClaimSpec{AccessModes:[ReadWriteOnce],Resources:ResourceRequirements{Limits:ResourceList{},Requests:ResourceList{storage: {{1073741824 0} {<nil>} 1Gi BinarySI},},},VolumeName:,Selector:nil,StorageClassName:*csi-rook-ceph-block,VolumeMode:*Filesystem,DataSource:nil,},Status:PersistentVolumeClaimStatus{Phase:,AccessModes:[],Capacity:ResourceList{},Conditions:[]PersistentVolumeClaimCondition{},},} 
I1117 06:07:39.838628      58 log.go:34] Create new csi pvc
Waiting up to &PersistentVolumeClaim{ObjectMeta:{rbd-pvc  default  11c4c863-b866-4cbc-8028-fe045f9e3bec  0 2021-11-17 06:06:11 +0000 UTC <nil> <nil> map[] map[] [] [kubernetes.io/pvc-protection]  [{kubectl-create Update v1 2021-11-17 06:06:11 +0000 UTC FieldsV1 {"f:spec":{"f:accessModes":{},"f:resources":{"f:requests":{".":{},"f:storage":{}}},"f:storageClassName":{},"f:volumeMode":{}}}} {kube-controller-manager Update v1 2021-11-17 06:06:27 +0000 UTC FieldsV1 {"f:metadata":{"f:annotations":{".":{},"f:pv.kubernetes.io/bind-completed":{},"f:pv.kubernetes.io/bound-by-controller":{},"f:volume.beta.kubernetes.io/storage-provisioner":{}}},"f:spec":{"f:volumeName":{}}}} {kube-controller-manager Update v1 2021-11-17 06:06:27 +0000 UTC FieldsV1 {"f:status":{"f:accessModes":{},"f:capacity":{".":{},"f:storage":{}},"f:phase":{}}}}]},Spec:PersistentVolumeClaimSpec{AccessModes:[ReadWriteOnce],Resources:ResourceRequirements{Limits:ResourceList{},Requests:ResourceList{storage: {{1073741824 0} {<nil>} 1Gi BinarySI},},},VolumeName:,Selector:nil,StorageClassName:*csi-rook-ceph-block,VolumeMode:*Filesystem,DataSource:nil,},Status:PersistentVolumeClaimStatus{Phase:,AccessModes:[],Capacity:ResourceList{},Conditions:[]PersistentVolumeClaimCondition{},},} to be in Bound state
waiting for PVC rbd-pvc (0 seconds elapsed) 
waiting for PVC rbd-pvc (2 seconds elapsed) 
I1117 06:07:41.870069      58 log.go:34] Waiting for PV pvc-4ae3e06d-3548-4350-b8e8-73232af13b7f to bind to PVC rbd-pvc
Waiting up to 30ms for PersistentVolumeClaims [rbd-pvc] to have phase Bound
PersistentVolumeClaim rbd-pvc found and phase=Bound (2.369103ms) 
Waiting up to 3s for PersistentVolume pvc-4ae3e06d-3548-4350-b8e8-73232af13b7f to have phase Bound 
PersistentVolume pvc-4ae3e06d-3548-4350-b8e8-73232af13b7f found and phase=Bound (2.895248ms)
I1117 06:07:41.885147      58 log.go:34] New PVC with same name "rbd-pvc" created via CSI
I1117 06:07:41.885173      58 log.go:34] Extracting new volume name from CSI PV
I1117 06:07:41.885178      58 log.go:34] CSI new volume name: csi-vol-ab608919-476c-11ec-a3f4-0242ac110005 
I1117 06:07:41.885181      58 log.go:34] Fetching csi pool name
I1117 06:07:41.885185      58 log.go:34] csi poolname: replicapool 
I1117 06:07:41.885189      58 log.go:34] Create new Ceph connection
I1117 06:07:41.885215      58 log.go:34] csi poolname: replicapool 
I1117 06:07:41.887651      58 log.go:34] clusterID: rook-ceph, monitors: 10.99.152.131:6789, poolname: replicapool
I1117 06:07:41.889883      58 log.go:34] New connection arg monitors: 10.99.152.131:6789, id: csi-rbd-provisioner, keyfile: /tmp/csi/keys/keyfile-223983377, pool: replicapool, datapool: 
I1117 06:07:41.889899      58 log.go:34] Cluster connection created
I1117 06:07:41.889903      58 log.go:34] Delete the placeholder CSI volume in ceph cluster
I1117 06:07:42.044115      58 log.go:34] Successfully removed volume csi-vol-ab608919-476c-11ec-a3f4-0242ac110005
I1117 06:07:42.044214      58 log.go:34] Rename old ceph volume to new CSI volume
I1117 06:07:42.107227      58 log.go:34] successfully renamed volume csi-vol-ab608919-476c-11ec-a3f4-0242ac110005 -> pvc-11c4c863-b866-4cbc-8028-fe045f9e3bec
I1117 06:07:42.107303      58 log.go:34] Delete old PV object: pvc-11c4c863-b866-4cbc-8028-fe045f9e3bec
waiting for PV pvc-11c4c863-b866-4cbc-8028-fe045f9e3bec in state &PersistentVolumeStatus{Phase:Bound,Message:,Reason:,} to be deleted (0 seconds elapsed) 
I1117 06:07:42.270506      58 log.go:34] deleted persistent volume &PersistentVolume{ObjectMeta:{pvc-11c4c863-b866-4cbc-8028-fe045f9e3bec    fcbeeb08-99f4-4ab1-bc75-52df73ed8e2a 1221 0 2021-11-17 06:06:27 +0000 UTC <nil> <nil> map[] map[pv.kubernetes.io/provisioned-by:ceph.rook.io/block] [] [kubernetes.io/pv-protection]  [{kube-controller-manager Update v1 2021-11-17 06:06:27 +0000 UTC FieldsV1 {"f:status":{"f:phase":{}}}} {rook Update v1 2021-11-17 06:06:27 +0000 UTC FieldsV1 {"f:metadata":{"f:annotations":{".":{},"f:pv.kubernetes.io/provisioned-by":{}}},"f:spec":{"f:accessModes":{},"f:capacity":{".":{},"f:storage":{}},"f:claimRef":{},"f:flexVolume":{".":{},"f:driver":{},"f:fsType":{},"f:options":{".":{},"f:clusterNamespace":{},"f:dataBlockPool":{},"f:image":{},"f:pool":{},"f:storageClass":{}}},"f:persistentVolumeReclaimPolicy":{},"f:storageClassName":{},"f:volumeMode":{}}}}]},Spec:PersistentVolumeSpec{Capacity:ResourceList{storage: {{1073741824 0} {<nil>} 1Gi BinarySI},},PersistentVolumeSource:PersistentVolumeSource{GCEPersistentDisk:nil,AWSElasticBlockStore:nil,HostPath:nil,Glusterfs:nil,NFS:nil,RBD:nil,ISCSI:nil,Cinder:nil,CephFS:nil,FC:nil,Flocker:nil,FlexVolume:&FlexPersistentVolumeSource{Driver:ceph.rook.io/rook-ceph,FSType:ext4,SecretRef:nil,ReadOnly:false,Options:map[string]string{clusterNamespace: rook-ceph,dataBlockPool: ,image: pvc-11c4c863-b866-4cbc-8028-fe045f9e3bec,pool: replicapool,storageClass: rook-ceph-block,},},AzureFile:nil,VsphereVolume:nil,Quobyte:nil,AzureDisk:nil,PhotonPersistentDisk:nil,PortworxVolume:nil,ScaleIO:nil,Local:nil,StorageOS:nil,CSI:nil,},AccessModes:[ReadWriteOnce],ClaimRef:&ObjectReference{Kind:PersistentVolumeClaim,Namespace:default,Name:rbd-pvc,UID:11c4c863-b866-4cbc-8028-fe045f9e3bec,APIVersion:v1,ResourceVersion:1140,FieldPath:,},PersistentVolumeReclaimPolicy:Retain,StorageClassName:rook-ceph-block,MountOptions:[],VolumeMode:*Filesystem,NodeAffinity:nil,},Status:PersistentVolumeStatus{Phase:Bound,Message:,Reason:,},}
I1117 06:07:42.270673      58 log.go:34] Successfully migrated all the PVCs to CSI```

Reply:
Thanks for sharing the log, here are a few final suggestions besides other places I commented:
- Inside DeletePVC(), change the logging so it just prints the name of the PVC, rather than the whole object
- This message is verbose: `Waiting up to &PersistentVolumeClaim...`
- There are some messages that don't have the timestamp. They need to use the new logger?

Reply:
Thanks for catching that! Actually, the logs that were not showing the timestamp with them were the ones that were printed by `fmt.Printf` and not the proper logger. Updated that now :+1: